File: cpp/benchmark/bench_mps_lu.cpp
// Two-phase Revised Simplex with hand-rolled LU factorization (no Eigen).
// Usage: ./bench_mps_lu <data.json> [N=10]

#include <algorithm>
#include <cassert>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <limits>
#include <numeric>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

// ---------------------------------------------------------------------------
// Minimal JSON parser (copied verbatim from bench_mps.cpp — no Eigen)
// ---------------------------------------------------------------------------

static std::string read_file(const std::string& path) {
    std::ifstream f(path);
    if (!f) throw std::runtime_error("Cannot open: " + path);
    std::ostringstream ss; ss << f.rdbuf(); return ss.str();
}
static int parse_int_key(const std::string& json, const std::string& key) {
    std::string pat = "\"" + key + "\"";
    std::size_t pos = json.find(pat);
    if (pos == std::string::npos) throw std::runtime_error("Key not found: " + key);
    pos = json.find(':', pos) + 1;
    while (pos < json.size() && (json[pos] == ' ' || json[pos] == '\t')) ++pos;
    return std::stoi(json.substr(pos));
}
static std::vector<double> parse_number_array(const std::string& json,
                                               std::size_t start, std::size_t& end) {
    std::vector<double> vals;
    if (json[start] != '[') throw std::runtime_error("Expected '['");
    std::size_t pos = start + 1;
    while (pos < json.size()) {
        char c = json[pos];
        if (c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == ',') { ++pos; continue; }
        if (c == ']') { end = pos + 1; return vals; }
        std::size_t num_end; double v = std::stod(json.substr(pos), &num_end);
        vals.push_back(v); pos += num_end;
    }
    throw std::runtime_error("Unterminated array");
}
static std::vector<std::vector<double>> parse_2d_array(const std::string& json,
                                                        std::size_t start, std::size_t& end) {
    std::vector<std::vector<double>> rows;
    if (json[start] != '[') throw std::runtime_error("Expected '['");
    std::size_t pos = start + 1;
    while (pos < json.size()) {
        char c = json[pos];
        if (c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == ',') { ++pos; continue; }
        if (c == ']') { end = pos + 1; return rows; }
        if (c == '[') {
            std::size_t re; rows.push_back(parse_number_array(json, pos, re)); pos = re; continue;
        }
        throw std::runtime_error("Unexpected char in 2D array");
    }
    throw std::runtime_error("Unterminated 2D array");
}
static std::size_t find_value_start(const std::string& json, const std::string& key) {
    std::string pat = "\"" + key + "\"";
    std::size_t pos = json.find(pat);
    if (pos == std::string::npos) throw std::runtime_error("Key not found: " + key);
    pos = json.find(':', pos) + 1;
    while (pos < json.size() && (json[pos]==' '||json[pos]=='\t'||json[pos]=='\n'||json[pos]=='\r'))
        ++pos;
    return pos;
}

struct LpData { int m, n; std::vector<double> A, b, c; };
static LpData load_lp(const std::string& path) {
    std::string json = read_file(path);
    LpData lp; lp.m = parse_int_key(json,"m"); lp.n = parse_int_key(json,"n");
    std::size_t a_start = find_value_start(json,"A"), a_end;
    auto rows = parse_2d_array(json, a_start, a_end);
    lp.A.resize(lp.m * lp.n);
    for (int i = 0; i < lp.m; ++i)
        for (int j = 0; j < lp.n; ++j) lp.A[i*lp.n+j] = rows[i][j];
    std::size_t b_start = find_value_start(json,"b"), b_end;
    lp.b = parse_number_array(json, b_start, b_end);
    std::size_t c_start = find_value_start(json,"c"), c_end;
    lp.c = parse_number_array(json, c_start, c_end);
    return lp;
}

// ---------------------------------------------------------------------------
// LU factorization (row-major flat vector)
// ---------------------------------------------------------------------------

struct LU {
    std::vector<double> m;
    std::vector<int> perm;
    int sz;
};

static LU lu_factor(const std::vector<double>& B, int sz) {
    LU lu; lu.sz = sz; lu.m = B;
    lu.perm.resize(sz); std::iota(lu.perm.begin(), lu.perm.end(), 0);
    for (int k = 0; k < sz; ++k) {
        int r = k; double maxv = std::abs(lu.m[k*sz+k]);
        for (int i = k+1; i < sz; ++i) {
            double v = std::abs(lu.m[i*sz+k]);
            if (v > maxv) { maxv = v; r = i; }
        }
        if (r != k) {
            for (int j = 0; j < sz; ++j) std::swap(lu.m[k*sz+j], lu.m[r*sz+j]);
            std::swap(lu.perm[k], lu.perm[r]);
        }
        if (std::abs(lu.m[k*sz+k]) < 1e-14) continue;
        for (int i = k+1; i < sz; ++i) {
            lu.m[i*sz+k] /= lu.m[k*sz+k];
            for (int j = k+1; j < sz; ++j)
                lu.m[i*sz+j] -= lu.m[i*sz+k] * lu.m[k*sz+j];
        }
    }
    return lu;
}

static std::vector<double> lu_solve_vec(const LU& lu, const std::vector<double>& b) {
    int sz = lu.sz;
    std::vector<double> y(sz);
    for (int i = 0; i < sz; ++i) y[i] = b[lu.perm[i]];
    for (int i = 1; i < sz; ++i)
        for (int k = 0; k < i; ++k) y[i] -= lu.m[i*sz+k]*y[k];
    std::vector<double> x = y;
    for (int i = sz-1; i >= 0; --i) {
        for (int k = i+1; k < sz; ++k) x[i] -= lu.m[i*sz+k]*x[k];
        if (std::abs(lu.m[i*sz+i]) > 1e-14) x[i] /= lu.m[i*sz+i];
    }
    return x;
}

// Solve B·X = N; N is m×nb row-major; returns m×nb row-major.
static std::vector<double> lu_solve_mat(const LU& lu, const std::vector<double>& N, int nb) {
    int m = lu.sz;
    std::vector<double> res(m*nb), col(m);
    for (int j = 0; j < nb; ++j) {
        for (int i = 0; i < m; ++i) col[i] = N[i*nb+j];
        auto x = lu_solve_vec(lu, col);
        for (int i = 0; i < m; ++i) res[i*nb+j] = x[i];
    }
    return res;
}

// ---------------------------------------------------------------------------
// Matrix helpers
// ---------------------------------------------------------------------------

static std::vector<double> cols_rm(const std::vector<double>& A, int m, int n,
                                    const std::vector<int>& idx) {
    int k = (int)idx.size();
    std::vector<double> out(m*k);
    for (int jj = 0; jj < k; ++jj) {
        int col = idx[jj];
        for (int i = 0; i < m; ++i) out[i*k+jj] = A[i*n+col];
    }
    return out;
}

static double vdot(const std::vector<double>& a, const std::vector<double>& b) {
    double s = 0; for (int i = 0; i < (int)a.size(); ++i) s += a[i]*b[i]; return s;
}

static std::vector<double> gather(const std::vector<double>& a, const std::vector<int>& idx) {
    std::vector<double> out(idx.size());
    for (int i = 0; i < (int)idx.size(); ++i) out[i] = a[idx[i]];
    return out;
}

// cr = cN - BinvN'·cB; BinvN is m×nb row-major
static std::vector<double> reduced_costs(const std::vector<double>& cN,
                                          const std::vector<double>& cB,
                                          const std::vector<double>& BinvN, int m, int nb) {
    std::vector<double> cr = cN;
    for (int j = 0; j < nb; ++j)
        for (int i = 0; i < m; ++i) cr[j] -= BinvN[i*nb+j]*cB[i];
    return cr;
}

static std::vector<double> dense_col(const std::vector<double>& BinvN, int m, int nb, int j) {
    std::vector<double> col(m);
    for (int i = 0; i < m; ++i) col[i] = BinvN[i*nb+j];
    return col;
}

// ---------------------------------------------------------------------------
// Pivot selection
// ---------------------------------------------------------------------------

static int entering_index(const std::vector<double>& cr, double tol) {
    int best = -1; double bv = tol;
    for (int j = 0; j < (int)cr.size(); ++j)
        if (cr[j] > bv) { bv = cr[j]; best = j; }
    return best;
}

static int leaving_index(const std::vector<double>& xb, const std::vector<double>& d,
                          const std::vector<double>& BinvN, int m, int nb, double tol) {
    std::vector<int> cands;
    for (int i = 0; i < m; ++i) if (d[i] > tol) cands.push_back(i);
    if (cands.empty()) return -1;
    int best = cands[0];
    for (int ii = 1; ii < (int)cands.size(); ++ii) {
        int i = cands[ii];
        double ri = xb[i]/d[i], rb = xb[best]/d[best];
        if (ri < rb - tol) { best = i; }
        else if (std::abs(ri-rb) <= tol) {
            for (int k = 0; k < nb; ++k) {
                double vi = BinvN[i*nb+k]/d[i], vb = BinvN[best*nb+k]/d[best];
                if (vi < vb-1e-14) { best = i; break; }
                else if (vi > vb+1e-14) break;
            }
        }
    }
    return best;
}

// ---------------------------------------------------------------------------
// Phase 1
// ---------------------------------------------------------------------------

static void make_b_nonneg(std::vector<double>& A, std::vector<double>& b, int m, int n) {
    for (int i = 0; i < m; ++i) if (b[i] < 0) {
        b[i] = -b[i];
        for (int j = 0; j < n; ++j) A[i*n+j] = -A[i*n+j];
    }
}

struct BFS { std::vector<int> basic, nonbasic; bool ok; };

static void fix_arts(std::vector<double>& Aw, std::vector<int>& basic,
                     std::vector<int>& nonbasic, int m, int nw, int artStart) {
    for (bool changed = true; changed;) {
        changed = false;
        for (int pos = 0; pos < (int)basic.size(); ++pos) {
            if (basic[pos] < artStart) continue;
            auto B = cols_rm(Aw, m, nw, basic);
            auto lu = lu_factor(B, m);
            auto BinvAw = lu_solve_mat(lu, Aw, nw);
            int swapJ = -1;
            for (int j = 0; j < (int)nonbasic.size(); ++j)
                if (nonbasic[j] < artStart && std::abs(BinvAw[pos*nw+nonbasic[j]]) > 1e-10)
                    { swapJ = j; break; }
            if (swapJ < 0) {
                for (int j = 0; j < nw; ++j) Aw[pos*nw+j] = 0.0;
            } else {
                std::swap(basic[pos], nonbasic[swapJ]);
                changed = true; break;
            }
        }
    }
}

static BFS phase1(const std::vector<double>& A_orig, const std::vector<double>& b_orig,
                  int m, int n, int maxit=10000, double tol=1e-8) {
    std::vector<double> A = A_orig, b = b_orig;
    make_b_nonneg(A, b, m, n);
    int nw = n + m;
    std::vector<double> Aw(m*nw, 0.0);
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) Aw[i*nw+j] = A[i*n+j];
        Aw[i*nw+n+i] = 1.0;
    }
    std::vector<double> cw(nw, 0.0);
    for (int i = n; i < nw; ++i) cw[i] = -1.0;

    std::vector<int> basic(m), nonbasic(n);
    std::iota(basic.begin(), basic.end(), n);
    std::iota(nonbasic.begin(), nonbasic.end(), 0);

    for (int it = 0; it < maxit; ++it) {
        int nb = (int)nonbasic.size();
        auto B = cols_rm(Aw, m, nw, basic);
        auto N = cols_rm(Aw, m, nw, nonbasic);
        auto lu = lu_factor(B, m);
        auto xb = lu_solve_vec(lu, b);
        auto BinvN = lu_solve_mat(lu, N, nb);
        auto cb = gather(cw, basic);
        auto cr = reduced_costs(gather(cw, nonbasic), cb, BinvN, m, nb);
        double z = vdot(cb, xb);

        int j = entering_index(cr, tol);
        if (j < 0) {
            if (z < -tol) return {{},{},false};
            fix_arts(Aw, basic, nonbasic, m, nw, n);
            std::vector<int> bas, nonbas;
            for (int v : basic)   if (v < n) bas.push_back(v);
            for (int v : nonbasic) if (v < n) nonbas.push_back(v);
            return {bas, nonbas, true};
        }
        auto d = dense_col(BinvN, m, nb, j);
        int i = leaving_index(xb, d, BinvN, m, nb, tol);
        if (i < 0) throw std::runtime_error("Phase 1 unexpectedly unbounded");
        std::swap(basic[i], nonbasic[j]);
    }
    throw std::runtime_error("Phase 1 did not converge");
}

// ---------------------------------------------------------------------------
// Phase 2
// ---------------------------------------------------------------------------

static std::pair<double,int> phase2(const std::vector<double>& A, const std::vector<double>& b,
                                     const std::vector<double>& c, int m, int n,
                                     std::vector<int> basic, std::vector<int> nonbasic,
                                     int maxit=10000, double tol=1e-10) {
    for (int it = 1; it <= maxit; ++it) {
        int nb = (int)nonbasic.size();
        auto B = cols_rm(A, m, n, basic);
        auto N = cols_rm(A, m, n, nonbasic);
        auto lu = lu_factor(B, m);
        auto xb = lu_solve_vec(lu, b);
        auto BinvN = lu_solve_mat(lu, N, nb);
        auto cb = gather(c, basic);
        auto cr = reduced_costs(gather(c, nonbasic), cb, BinvN, m, nb);
        double z = vdot(cb, xb);

        int j = entering_index(cr, tol);
        if (j < 0) return {z, it};
        auto d = dense_col(BinvN, m, nb, j);
        int i = leaving_index(xb, d, BinvN, m, nb, tol);
        if (i < 0) return {std::numeric_limits<double>::infinity(), it};
        std::swap(basic[i], nonbasic[j]);
    }
    throw std::runtime_error("Phase 2 did not converge");
}

static std::pair<double,int> simplex_solve(const LpData& lp) {
    int m = lp.m, n = lp.n;
    auto bfs = phase1(lp.A, lp.b, m, n);
    if (!bfs.ok) return {0.0, 0};
    // fill any unaccounted columns into nonbasic
    std::unordered_map<int,bool> inSet;
    for (int v : bfs.basic)    inSet[v] = true;
    for (int v : bfs.nonbasic) inSet[v] = true;
    for (int j = 0; j < n; ++j)
        if (!inSet[j]) bfs.nonbasic.push_back(j);
    return phase2(lp.A, lp.b, lp.c, m, n, bfs.basic, bfs.nonbasic);
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------

int main(int argc, char* argv[]) {
    if (argc < 2) { std::cerr << "Usage: bench_mps_lu <data.json> [N]\n"; return 1; }
    int n_runs = (argc >= 3) ? std::atoi(argv[2]) : 10;
    LpData lp = load_lp(argv[1]);

    using Clock = std::chrono::high_resolution_clock;
    using Dur   = std::chrono::duration<double, std::milli>;

    double last_z = 0.0; int last_it = 0;
    std::vector<double> times_ms; times_ms.reserve(n_runs);

    for (int r = 0; r < n_runs; ++r) {
        auto t0 = Clock::now();
        auto [z, it] = simplex_solve(lp);
        times_ms.push_back(Dur(Clock::now() - t0).count());
        last_z = z; last_it = it;
    }

    std::cout << "{\"z\": " << std::fixed; std::cout.precision(4);
    std::cout << last_z << ", \"iterations\": " << last_it << ", \"times_ms\": [";
    std::cout.precision(6);
    for (int i = 0; i < n_runs; ++i) { if (i) std::cout << ", "; std::cout << times_ms[i]; }
    std::cout << "]}\n";
    return 0;
}

File: rust/src/bin/bench_mps_lu.rs
// Two-phase Revised Simplex with hand-rolled LU factorization (no nalgebra).
// Usage: ./bench_mps_lu <data.json> [N=10]

use serde::Deserialize;
use std::fs;
use std::time::Instant;

#[derive(Deserialize)]
struct LpData {
    #[allow(dead_code)]
    name: String,
    m: usize,
    n: usize,
    #[serde(rename = "A")]
    a: Vec<Vec<f64>>,
    b: Vec<f64>,
    c: Vec<f64>,
}

// ---------------------------------------------------------------------------
// LU factorization with partial pivoting — flat row-major Vec<f64>
// ---------------------------------------------------------------------------

struct LuFact {
    m: Vec<f64>,   // packed LU sz×sz row-major
    perm: Vec<usize>,
    sz: usize,
}

fn lu_factor(b: &[f64], sz: usize) -> LuFact {
    let mut m = b.to_vec();
    let mut perm: Vec<usize> = (0..sz).collect();
    for k in 0..sz {
        let (mut r, mut maxv) = (k, m[k * sz + k].abs());
        for i in k + 1..sz {
            let v = m[i * sz + k].abs();
            if v > maxv {
                maxv = v;
                r = i;
            }
        }
        if r != k {
            for j in 0..sz {
                m.swap(k * sz + j, r * sz + j);
            }
            perm.swap(k, r);
        }
        if m[k * sz + k].abs() < 1e-14 {
            continue;
        }
        for i in k + 1..sz {
            m[i * sz + k] /= m[k * sz + k];
            for j in k + 1..sz {
                let mul = m[i * sz + k];
                m[i * sz + j] -= mul * m[k * sz + j];
            }
        }
    }
    LuFact { m, perm, sz }
}

impl LuFact {
    fn solve_vec(&self, b: &[f64]) -> Vec<f64> {
        let sz = self.sz;
        let mut y: Vec<f64> = self.perm.iter().map(|&i| b[i]).collect();
        for i in 1..sz {
            for k in 0..i {
                let v = self.m[i * sz + k] * y[k];
                y[i] -= v;
            }
        }
        let mut x = y;
        for i in (0..sz).rev() {
            for k in i + 1..sz {
                let v = self.m[i * sz + k] * x[k];
                x[i] -= v;
            }
            let d = self.m[i * sz + i];
            if d.abs() > 1e-14 {
                x[i] /= d;
            }
        }
        x
    }

    // N is m×nb row-major; returns m×nb row-major.
    fn solve_mat(&self, n_mat: &[f64], nb: usize) -> Vec<f64> {
        let m = self.sz;
        let mut res = vec![0f64; m * nb];
        let mut col = vec![0f64; m];
        for j in 0..nb {
            for i in 0..m {
                col[i] = n_mat[i * nb + j];
            }
            let x = self.solve_vec(&col);
            for i in 0..m {
                res[i * nb + j] = x[i];
            }
        }
        res
    }
}

// ---------------------------------------------------------------------------
// Matrix helpers
// ---------------------------------------------------------------------------

fn cols_rm(a: &[f64], m: usize, n: usize, idx: &[usize]) -> Vec<f64> {
    let k = idx.len();
    let mut out = vec![0f64; m * k];
    for (jj, &col) in idx.iter().enumerate() {
        for i in 0..m {
            out[i * k + jj] = a[i * n + col];
        }
    }
    out
}

fn dot(a: &[f64], b: &[f64]) -> f64 {
    a.iter().zip(b).map(|(x, y)| x * y).sum()
}

fn gather(a: &[f64], idx: &[usize]) -> Vec<f64> {
    idx.iter().map(|&i| a[i]).collect()
}

fn reduced_costs(c_n: &[f64], c_b: &[f64], binv_n: &[f64], m: usize, nb: usize) -> Vec<f64> {
    let mut cr = c_n.to_vec();
    for j in 0..nb {
        for i in 0..m {
            cr[j] -= binv_n[i * nb + j] * c_b[i];
        }
    }
    cr
}

fn dense_col(binv_n: &[f64], m: usize, nb: usize, j: usize) -> Vec<f64> {
    (0..m).map(|i| binv_n[i * nb + j]).collect()
}

// ---------------------------------------------------------------------------
// Pivot selection
// ---------------------------------------------------------------------------

fn entering_index(cr: &[f64], tol: f64) -> Option<usize> {
    let (mut best, mut best_val) = (usize::MAX, tol);
    for (j, &v) in cr.iter().enumerate() {
        if v > best_val {
            best_val = v;
            best = j;
        }
    }
    if best == usize::MAX { None } else { Some(best) }
}

fn leaving_index(xb: &[f64], d: &[f64], binv_n: &[f64], m: usize, nb: usize, tol: f64) -> Option<usize> {
    let cands: Vec<usize> = (0..m).filter(|&i| d[i] > tol).collect();
    if cands.is_empty() {
        return None;
    }
    let mut best = cands[0];
    for &i in &cands[1..] {
        let ri = xb[i] / d[i];
        let rb = xb[best] / d[best];
        if ri < rb - tol {
            best = i;
        } else if (ri - rb).abs() <= tol {
            for k in 0..nb {
                let vi = binv_n[i * nb + k] / d[i];
                let vb = binv_n[best * nb + k] / d[best];
                if vi < vb - 1e-14 {
                    best = i;
                    break;
                } else if vi > vb + 1e-14 {
                    break;
                }
            }
        }
    }
    Some(best)
}

// ---------------------------------------------------------------------------
// Phase 1
// ---------------------------------------------------------------------------

fn make_b_nonneg(a: &mut Vec<f64>, b: &mut Vec<f64>, m: usize, n: usize) {
    for i in 0..m {
        if b[i] < 0.0 {
            b[i] = -b[i];
            for j in 0..n {
                a[i * n + j] = -a[i * n + j];
            }
        }
    }
}

fn phase1(
    a_orig: &[f64], b_orig: &[f64], m: usize, n: usize,
    maxit: usize, tol: f64,
) -> Option<(Vec<usize>, Vec<usize>)> {
    let mut a = a_orig.to_vec();
    let mut b = b_orig.to_vec();
    make_b_nonneg(&mut a, &mut b, m, n);

    let nw = n + m;
    let mut aw = vec![0f64; m * nw];
    for i in 0..m {
        for j in 0..n {
            aw[i * nw + j] = a[i * n + j];
        }
        aw[i * nw + n + i] = 1.0;
    }
    let mut cw = vec![0f64; nw];
    for i in n..nw {
        cw[i] = -1.0;
    }

    let mut basic: Vec<usize> = (n..n + m).collect();
    let mut nonbasic: Vec<usize> = (0..n).collect();

    for _ in 0..maxit {
        let nb = nonbasic.len();
        let b_mat = cols_rm(&aw, m, nw, &basic);
        let n_mat = cols_rm(&aw, m, nw, &nonbasic);
        let lu = lu_factor(&b_mat, m);
        let xb = lu.solve_vec(&b);
        let binv_n = lu.solve_mat(&n_mat, nb);
        let cb = gather(&cw, &basic);
        let cn = gather(&cw, &nonbasic);
        let cr = reduced_costs(&cn, &cb, &binv_n, m, nb);
        let z = dot(&cb, &xb);

        match entering_index(&cr, tol) {
            None => {
                if z < -tol {
                    return None;
                }
                fix_arts(&mut aw, &mut basic, &mut nonbasic, m, nw, n);
                let bas: Vec<usize> = basic.iter().copied().filter(|&v| v < n).collect();
                let nonbas: Vec<usize> = nonbasic.iter().copied().filter(|&v| v < n).collect();
                return Some((bas, nonbas));
            }
            Some(j) => {
                let d = dense_col(&binv_n, m, nb, j);
                match leaving_index(&xb, &d, &binv_n, m, nb, tol) {
                    None => panic!("Phase 1 unexpectedly unbounded"),
                    Some(i) => {
                        basic.swap(i, i); // keep borrow checker happy
                        let tmp = basic[i];
                        basic[i] = nonbasic[j];
                        nonbasic[j] = tmp;
                    }
                }
            }
        }
    }
    panic!("Phase 1 did not converge");
}

fn fix_arts(aw: &mut Vec<f64>, basic: &mut Vec<usize>, nonbasic: &mut Vec<usize>,
            m: usize, nw: usize, art_start: usize) {
    loop {
        let mut changed = false;
        for pos in 0..basic.len() {
            if basic[pos] < art_start {
                continue;
            }
            let b_mat = cols_rm(aw, m, nw, basic);
            let lu = lu_factor(&b_mat, m);
            let binv_aw = lu.solve_mat(aw, nw);
            let swap_j = nonbasic.iter().position(|&nc| {
                nc < art_start && binv_aw[pos * nw + nc].abs() > 1e-10
            });
            match swap_j {
                None => {
                    for j in 0..nw {
                        aw[pos * nw + j] = 0.0;
                    }
                }
                Some(j) => {
                    let tmp = basic[pos];
                    basic[pos] = nonbasic[j];
                    nonbasic[j] = tmp;
                    changed = true;
                    break;
                }
            }
        }
        if !changed {
            break;
        }
    }
}

// ---------------------------------------------------------------------------
// Phase 2
// ---------------------------------------------------------------------------

fn phase2(
    a: &[f64], b: &[f64], c: &[f64], m: usize, n: usize,
    basic_in: &[usize], nonbasic_in: &[usize],
    maxit: usize, tol: f64,
) -> (f64, usize, bool) {
    let mut basic = basic_in.to_vec();
    let mut nonbasic = nonbasic_in.to_vec();

    for it in 1..=maxit {
        let nb = nonbasic.len();
        let b_mat = cols_rm(a, m, n, &basic);
        let n_mat = cols_rm(a, m, n, &nonbasic);
        let lu = lu_factor(&b_mat, m);
        let xb = lu.solve_vec(b);
        let binv_n = lu.solve_mat(&n_mat, nb);
        let cb = gather(c, &basic);
        let cn = gather(c, &nonbasic);
        let cr = reduced_costs(&cn, &cb, &binv_n, m, nb);
        let z = dot(&cb, &xb);

        match entering_index(&cr, tol) {
            None => return (z, it, true),
            Some(j) => {
                let d = dense_col(&binv_n, m, nb, j);
                match leaving_index(&xb, &d, &binv_n, m, nb, tol) {
                    None => return (f64::INFINITY, it, false),
                    Some(i) => {
                        let tmp = basic[i];
                        basic[i] = nonbasic[j];
                        nonbasic[j] = tmp;
                    }
                }
            }
        }
    }
    panic!("Phase 2 did not converge");
}

fn simplex_solve(a: &[f64], b: &[f64], c: &[f64], m: usize, n: usize) -> (f64, usize) {
    match phase1(a, b, m, n, 10000, 1e-8) {
        None => (0.0, 0),
        Some((basic, mut nonbasic)) => {
            let in_set: std::collections::HashSet<usize> =
                basic.iter().chain(nonbasic.iter()).copied().collect();
            for j in 0..n {
                if !in_set.contains(&j) {
                    nonbasic.push(j);
                }
            }
            let (z, it, _) = phase2(a, b, c, m, n, &basic, &nonbasic, 10000, 1e-10);
            (z, it)
        }
    }
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: bench_mps_lu <data.json> [N]");
        std::process::exit(1);
    }
    let n_runs: usize = if args.len() >= 3 { args[2].parse().unwrap_or(10) } else { 10 };

    let raw = fs::read_to_string(&args[1]).expect("Cannot read JSON file");
    let data: LpData = serde_json::from_str(&raw).expect("Cannot parse JSON");

    let m = data.m;
    let n = data.n;
    let a_flat: Vec<f64> = (0..m).flat_map(|i| data.a[i].iter().copied()).collect();
    let b = &data.b;
    let c = &data.c;

    let mut last_z = 0.0f64;
    let mut last_it = 0usize;
    let mut times_ms: Vec<f64> = Vec::with_capacity(n_runs);

    for _ in 0..n_runs {
        let t0 = Instant::now();
        let (z, it) = simplex_solve(&a_flat, b, c, m, n);
        times_ms.push(t0.elapsed().as_secs_f64() * 1000.0);
        last_z = z;
        last_it = it;
    }

    let times_str = times_ms
        .iter()
        .map(|t| format!("{:.6}", t))
        .collect::<Vec<_>>()
        .join(", ");

    println!(
        "{{\"z\": {:.4}, \"iterations\": {}, \"times_ms\": [{}]}}",
        last_z, last_it, times_str
    );
}
